File: registry-access/commands/src/deprecation/undeprecate.ts
import { docsUrl } from '@pnpm/cli.utils'
import { PnpmError } from '@pnpm/error'
import { renderHelp } from 'render-help'

import { cliOptionsTypes, type DeprecateOptions, parsePackageSpec, rcOptionsTypes, updateDeprecation } from './common.js'

export { cliOptionsTypes, rcOptionsTypes }

export const commandNames = ['undeprecate']

export function help (): string {
  return renderHelp({
    description: 'Removes deprecation from a version of a package in the registry. Only works on already deprecated versions.',
    descriptionLists: [
      {
        title: 'Options',

        list: [
          {
            description: 'The base URL of the npm registry.',
            name: '--registry <url>',
          },
          {
            description: 'When publishing packages that require two-factor authentication, this option can specify a one-time password.',
            name: '--otp',
          },
        ],
      },
    ],
    url: docsUrl('undeprecate'),
    usages: [
      'pnpm undeprecate <package>[@<version>]',
    ],
  })
}

export async function handler (
  opts: DeprecateOptions,
  params: string[]
): Promise<void> {
  if (params.length === 0) {
    throw new PnpmError('UNDEPRECATE_REQUIRED', 'Package name is required')
  }

  if (params.length > 1) {
    throw new PnpmError('UNDEPRECATE_NO_MESSAGE', 'The undeprecate command does not accept a message.')
  }

  const { name, version } = parsePackageSpec(params[0])

  await updateDeprecation(name, version, '', opts, true)
}

File: pnpm/crates/cli/src/cli_args/undeprecate.rs
use clap::Args;
use pacquet_config::Config;

use super::deprecate::{
    DeprecateContext, DeprecateError, PackageSpec, parse_package_spec, update_deprecation,
};

#[derive(Debug, Args)]
pub struct UndeprecateArgs {
    /// The base URL of the npm registry.
    #[clap(long)]
    pub registry: Option<String>,

    /// One-time password for registries that require two-factor authentication.
    #[clap(long)]
    pub otp: Option<String>,

    /// The package name.
    pub params: Vec<String>,
}

impl UndeprecateArgs {
    pub async fn run(self, config: &Config) -> miette::Result<Option<String>> {
        let context = DeprecateContext::new(config, self.registry.as_ref(), self.otp)?;

        let spec = self.params.first().ok_or(DeprecateError::PackageRequired)?;
        let PackageSpec { name: package_name, version } = parse_package_spec(spec)?;

        let output = update_deprecation(&context, None, &package_name, version.as_deref()).await?;
        Ok(Some(output))
    }
}
